pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/rest/api/FunctionsImpl.java

```diff
@@ -526,6 +528,84 @@ public Response triggerFunction(final @PathParam("tenant") String tenant,
         }
     }
 
+    @POST
+    @Path("/{tenant}/{namespace}/{functionName}/upload")
+    @Consumes(MediaType.MULTIPART_FORM_DATA)
+    public Response uploadFunction(final @PathParam("tenant") String tenant,
+                                   final @PathParam("namespace") String namespace,
+                                   final @PathParam("functionName") String functionName,
+                                   final @FormDataParam("data") InputStream uploadedInputStream,
+                                   final @FormDataParam("data") FormDataContentDisposition fileDetail) {
+        // validate parameters
+        try {
+            if (tenant == null) {
+                throw new IllegalArgumentException("Tenant is not provided");
+            }
+            if (namespace == null) {
+                throw new IllegalArgumentException("Namespace is not provided");
+            }
+            if (functionName == null) {
+                throw new IllegalArgumentException("Function Name is not provided");
+            }
+            if (uploadedInputStream == null || fileDetail == null) {
+                throw new IllegalArgumentException("Function Package is not provided");
+            }
+        } catch (IllegalArgumentException e) {
+            log.error("Invalid upload function request @ /{}/{}/{}",
+                    tenant, namespace, functionName, e);
+            return Response.status(Status.BAD_REQUEST)
+                    .type(MediaType.APPLICATION_JSON)
+                    .entity(new ErrorData(e.getMessage())).build();
+        }
+
+        String packageLocation = String.format(
+                        "%s/%s/%s/%s",
+                        tenant,
+                        namespace,
+                        functionName,
+                        fileDetail.getFileName());
+
+        // Upload to bookkeeper
+        try {
+            log.info("Uploading function package to {}", packageLocation);
+
+            Utils.uploadToBookeeper(
+                    worker().getDlogNamespace(),
+                    uploadedInputStream,
+                    packageLocation);
+        } catch (IOException e) {
+            log.error("Error uploading file {}", packageLocation, e);
+            return Response.serverError()
+                    .type(MediaType.APPLICATION_JSON)
+                    .entity(new ErrorData(e.getMessage()))
+                    .build();
+        }
+
+        return Response.status(Status.OK).build();
+    }
+
+    @GET
+    @Path("/{tenant}/{namespace}/{functionName}/download")
+    public Response downloadFunction(final @PathParam("tenant") String tenant,
+                                     final @PathParam("namespace") String namespace,
+                                     final @PathParam("functionName") String functionName,
+                                     final @QueryParam("filename") String fileName) {
+        String packageLocation = String.format(
```

Comment:
should we have a common function for this so we can standardize the format for the path? Since we have the same thing in registerFunction.

Reply:
Actually there are changes between the way this uses and the way create works. That's why I didn't bother refactoring